api/pkg/authz/enforcer/enforcer.go

```diff
@@ -114,244 +71,207 @@ func newEnforcer(
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to get user roles: %w", err)
+	}
+	roles := make([]string, 0)
+	for _, tuple := range roleRelationships.RelationTuples {
+		roles = append(roles, tuple.Object)
 	}
-	payload := res.GetPayload()
-	return &types.Policy{
-		ID:        payload.ID,
-		Actions:   payload.Actions,
-		Resources: payload.Resources,
-		Subjects:  payload.Subjects,
-	}, nil
-}
-
-// FilterAuthorizedResource filter and return list of authorized resource for certain user
-func (e *enforcer) FilterAuthorizedResource(user string, resources []string, action string) ([]string, error) {
-	user = e.formatUser(user)
 
-	var wg sync.WaitGroup
-	allowedResourcesConcurrent := util.ConcurrentSlice{}
-	errorsConcurrent := util.ConcurrentSlice{}
-	for _, resource := range resources {
-		wg.Add(1)
-		go func(u string, r string, a string) {
-			defer wg.Done()
-			r = e.formatResource(r)
+	return roles, nil
+}
 
-			allowed, err := e.isAllowed(u, r, a)
-			if err != nil {
-				errorsConcurrent.Append(err)
-				return
-			}
-			if *allowed {
-				allowedResourcesConcurrent.Append(e.stripResourcePrefix(r))
-			}
-		}(user, resource, action)
+func (e *enforcer) GetRolePermissions(ctx context.Context, role string) ([]string, error) {
+	permissionRelationships, _, err := e.ketoReadClient.RelationshipApi.GetRelationships(ctx).
+		Namespace("Permission").
+		Relation("granted").
+		SubjectSetNamespace("Role").
+		SubjectSetRelation("member").
+		SubjectSetObject(role).Execute()
+	if err != nil {
+		return nil, err
 	}
-	wg.Wait()
-
-	errors := errorsConcurrent.GetItems()
-	if len(errors) > 0 {
-		return nil, errors[0].(error)
+	permissions := make([]string, 0)
+	for _, tuple := range permissionRelationships.RelationTuples {
+		permissions = append(permissions, tuple.Object)
 	}
 
-	allowedResources := make([]string, 0)
-	for _, item := range allowedResourcesConcurrent.GetItems() {
-		allowedResources = append(allowedResources, item.(string))
-	}
-	sort.Strings(allowedResources)
-	return allowedResources, nil
+	return permissions, nil
 }
 
-// UpsertRole create or update a role containing member as specified by users argument
-func (e *enforcer) UpsertRole(roleName string, users []string) (*types.Role, error) {
-	fmtRoleName := e.formatRole(roleName)
-	fmtUser := make([]string, 0)
-	for _, user := range users {
-		fmtUser = append(fmtUser, e.formatUser(user))
-	}
-
-	input := &models.OryAccessControlPolicyRole{
-		ID:      fmtRoleName,
-		Members: fmtUser,
-	}
-	params := &engines.UpsertOryAccessControlPolicyRoleParams{
-		Body:   input,
-		Flavor: string(e.flavor),
-	}
-	res, err := e.ketoClient.UpsertOryAccessControlPolicyRole(params.WithTimeout(e.timeout))
+func (e *enforcer) GetRoleMembers(ctx context.Context, role string) ([]string, error) {
+	expandedRole, _, err := e.ketoReadClient.PermissionApi.ExpandPermissions(ctx).
+		Namespace("Role").
+		Relation("member").
+		Object(role).
+		MaxDepth(2).Execute()
```

Comment:
Keto 0.6+ store the relation as a graph, and the root node in this case is a node named "Role:role", and the child nodes are the role members. Max depth of 1 will mean that we only get the root node, based on my test. Though i will check this again to be sure.

As for why we don't go beyond 2 - that's because we have not implemented role inheritance at the moment: i.e. we haven't done something like, admin role member is also a reader role, so the role member relationship tree will have depth 2 at most.

Reply:
I see.. this is useful information. Can you help capture this as an inline comment, for future reference?